numpy/core/include/numpy/npy_math.h

```diff
@@ -360,84 +360,121 @@ NPY_INPLACE npy_longdouble npy_heavisidel(npy_longdouble x, npy_longdouble h0);
  * Complex declarations
  */
 
-/*
- * C99 specifies that complex numbers have the same representation as
- * an array of two elements, where the first element is the real part
- * and the second element is the imaginary part.
- */
-#define __NPY_CPACK_IMP(x, y, type, ctype)   \
-    union {                                  \
-        ctype z;                             \
-        type a[2];                           \
-    } z1;                                    \
-                                             \
-    z1.a[0] = (x);                           \
-    z1.a[1] = (y);                           \
-                                             \
-    return z1.z;
+#define REAL 0
+#define IMAG 1
+#define CREATE_UNION(ctype, npy_ctype)  \
+    typedef union {                     \
+        ctype *arr;                     \
+        const npy_ctype *comp;          \
+    } _u;
+#define NPY_COMPLEX_GET(ctype, npy_ctype, c, real_or_imag)  \
+    CREATE_UNION(ctype, npy_ctype)                          \
+    _u tmp;                                                 \
+    tmp.comp = &c;                                           \
+    return tmp.arr[real_or_imag];                           \
+
+#define NPY_COMPLEX_SET(ctype, npy_ctype, c, val, real_or_imag) \
+    CREATE_UNION(ctype, npy_ctype)                              \
+    _u tmp;                                                     \
+    tmp.comp = c;                                               \
+    tmp.arr[real_or_imag] = val;                                \
+
+static inline double npy_creal(const npy_cdouble z)
+{
+    NPY_COMPLEX_GET(double, npy_cdouble, z, REAL)
+}
 
-static inline npy_cdouble npy_cpack(double x, double y)
+static inline void npy_csetreal(npy_cdouble *z, const double r)
 {
-    __NPY_CPACK_IMP(x, y, double, npy_cdouble);
+    NPY_COMPLEX_SET(double, npy_cdouble, z, r, REAL)
 }
 
-static inline npy_cfloat npy_cpackf(float x, float y)
+static inline double npy_cimag(const npy_cdouble z)
 {
-    __NPY_CPACK_IMP(x, y, float, npy_cfloat);
+    NPY_COMPLEX_GET(double, npy_cdouble, z, IMAG)
 }
 
-static inline npy_clongdouble npy_cpackl(npy_longdouble x, npy_longdouble y)
+static inline void npy_csetimag(npy_cdouble *z, const double i)
 {
-    __NPY_CPACK_IMP(x, y, npy_longdouble, npy_clongdouble);
+    NPY_COMPLEX_SET(double, npy_cdouble, z, i, IMAG)
```

Comment:
I remember having trouble with doing this when I started working on this PR, but things have changed since then and it appears that all is okay to use it now. Change pushed.